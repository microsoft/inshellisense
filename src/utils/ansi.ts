// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

const ESC = "\u001B";
const CSI = ESC + "[";
const OSC = "\u001B]";
const BEL = "\u0007";

export const IsTermOscPs = 6973;
const IS_OSC = OSC + IsTermOscPs + ";";

export enum IstermOscPt {
  PromptStarted = "PS",
  PromptEnded = "PE",
  CurrentWorkingDirectory = "CWD",
}

export const IstermPromptStart = IS_OSC + IstermOscPt.PromptStarted + BEL;
export const IstermPromptEnd = IS_OSC + IstermOscPt.PromptEnded + BEL;
export const cursorHide = CSI + "?25l";
export const cursorShow = CSI + "?25h";
export const cursorNextLine = CSI + "E";
export const eraseLine = CSI + "2K";
export const cursorBackward = (count = 1) => CSI + count + "D";
export const cursorTo = ({ x, y }: { x?: number; y?: number }) => {
  return CSI + (y ?? "") + ";" + (x ?? "") + "H";
};
export const deleteLinesBelow = (count = 1) => {
  return [...Array(count).keys()].map(() => CSI + "B" + CSI + "M").join("");
};
export const deleteLine = (count = 1) => CSI + count + "M";
export const scrollUp = (count = 1) => CSI + count + "S";
export const scrollDown = (count = 1) => CSI + count + "T";
export const eraseLinesBelow = (count = 1) => {
  return [...Array(count).keys()].map(() => cursorNextLine + eraseLine).join("");
};
