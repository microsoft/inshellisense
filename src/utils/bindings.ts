// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import os from "node:os";
import path from "node:path";
import fsAsync from "node:fs/promises";
import fs from "node:fs";
import process from "node:process";
import url from "node:url";
import { exec } from "node:child_process";
import util from "node:util";
const execAsync = util.promisify(exec);

const __filename = url.fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const cacheFolder = ".inshellisense";

export enum Shell {
  Bash = "bash",
  Powershell = "powershell",
  Pwsh = "pwsh",
  Zsh = "zsh",
  Fish = "fish",
}

export const supportedShells = [Shell.Bash, process.platform == "win32" ? Shell.Powershell : null, Shell.Pwsh, Shell.Zsh, Shell.Fish].filter(
  (shell) => shell != null,
) as Shell[];

const bashScriptCommand = (): string => {
  return `[ -f ~/${cacheFolder}/key-bindings.bash ] && source ~/${cacheFolder}/key-bindings.bash`;
};

const zshScriptCommand = (): string => {
  return `[ -f ~/${cacheFolder}/key-bindings.zsh ] && source ~/${cacheFolder}/key-bindings.zsh`;
};

const fishScriptCommand = (): string => {
  return `[ -f ~/${cacheFolder}/key-bindings.fish ] && source ~/${cacheFolder}/key-bindings.fish`;
};

const powershellScriptCommand = (): string => {
  const bindingsPath = path.join(os.homedir(), cacheFolder, "key-bindings-powershell.ps1");
  return `if(Test-Path '${bindingsPath}' -PathType Leaf){. ${bindingsPath}}`;
};

const pwshScriptCommand = (): string => {
  const bindingsPath = path.join(os.homedir(), cacheFolder, "key-bindings-pwsh.ps1");
  return `if(Test-Path '${bindingsPath}' -PathType Leaf){. ${bindingsPath}}`;
};

const pwshConfigPath = async (): Promise<string> => {
  try {
    const { stdout } = await execAsync("echo $profile", { shell: "pwsh" });
    return stdout.trim();
  } catch {
    /* empty */
  }
  switch (process.platform) {
    case "win32":
      return path.join(os.homedir(), "Documents", "Powershell", "Microsoft.PowerShell_profile.ps1");
    case "linux":
    case "darwin":
      return path.join(os.homedir(), ".config", "powershell", "Microsoft.PowerShell_profile.ps1");
    default:
      throw new Error("Unsupported platform");
  }
};

const powershellConfigPath = async (): Promise<string> => {
  try {
    const { stdout } = await execAsync("echo $profile", { shell: "powershell" });
    return stdout.trim();
  } catch {
    /* empty */
  }
  return path.join(os.homedir(), "Documents", "WindowsPowershell", "Microsoft.PowerShell_profile.ps1");
};

export const availableBindings = async (): Promise<Shell[]> => {
  const cliConfigPath = path.join(os.homedir(), cacheFolder);
  if (!fs.existsSync(cliConfigPath)) {
    await fsAsync.mkdir(cliConfigPath);
  }

  const bindings = [];
  const bashConfigPath = path.join(os.homedir(), ".bashrc");
  if (!fs.existsSync(bashConfigPath)) {
    bindings.push(Shell.Bash);
  } else {
    const bashConfigContent = fsAsync.readFile(bashConfigPath, { encoding: "utf-8" });
    if (!(await bashConfigContent).includes(bashScriptCommand())) {
      bindings.push(Shell.Bash);
    }
  }

  const zshConfigPath = path.join(os.homedir(), ".zshrc");
  if (!fs.existsSync(zshConfigPath)) {
    bindings.push(Shell.Zsh);
  } else {
    const zshConfigContent = fsAsync.readFile(zshConfigPath, { encoding: "utf-8" });
    if (!(await zshConfigContent).includes(zshScriptCommand())) {
      bindings.push(Shell.Zsh);
    }
  }

  const fishConfigPath = path.join(os.homedir(), ".config", "fish", "config.fish");
  if (!fs.existsSync(fishConfigPath)) {
    bindings.push(Shell.Fish);
  } else {
    const fishConfigContent = fsAsync.readFile(fishConfigPath, { encoding: "utf-8" });
    if (!(await fishConfigContent).includes(fishScriptCommand())) {
      bindings.push(Shell.Fish);
    }
  }

  if (process.platform == "win32") {
    const powershellResolvedConfigPath = await powershellConfigPath();
    if (!fs.existsSync(powershellResolvedConfigPath)) {
      bindings.push(Shell.Powershell);
    } else {
      const powershellConfigContent = fsAsync.readFile(powershellResolvedConfigPath, { encoding: "utf-8" });
      if (!(await powershellConfigContent).includes(powershellScriptCommand())) {
        bindings.push(Shell.Powershell);
      }
    }
  }

  const pwshResolvedConfigPath = await pwshConfigPath();
  if (!fs.existsSync(pwshResolvedConfigPath)) {
    bindings.push(Shell.Pwsh);
  } else {
    const pwshConfigContent = fsAsync.readFile(pwshResolvedConfigPath, { encoding: "utf-8" });
    if (!(await pwshConfigContent).includes(pwshScriptCommand())) {
      bindings.push(Shell.Pwsh);
    }
  }

  return bindings;
};

export const unbindAll = async (): Promise<void> => {
  try {
    const bashConfigPath = path.join(os.homedir(), ".bashrc");
    const bashConfig = (await fsAsync.readFile(bashConfigPath)).toString();
    if (bashConfig.includes(bashScriptCommand())) {
      const unboundBashConfig = bashConfig.toString().replace(bashScriptCommand(), "");
      await fsAsync.writeFile(bashConfigPath, unboundBashConfig);
    }
  } catch {
    /* empty */
  }

  try {
    const zshConfigPath = path.join(os.homedir(), ".zshrc");
    const zshConfig = (await fsAsync.readFile(zshConfigPath)).toString();
    if (zshConfig.includes(zshScriptCommand())) {
      const unboundZshConfig = zshConfig.toString().replace(zshScriptCommand(), "");
      await fsAsync.writeFile(zshConfigPath, unboundZshConfig);
    }
  } catch {
    /* empty */
  }

  try {
    const fishConfigPath = path.join(os.homedir(), ".config", "fish", "config.fish");
    const fishConfig = (await fsAsync.readFile(fishConfigPath)).toString();
    if (fishConfig.includes(fishScriptCommand())) {
      const unboundFishConfig = fishConfig.toString().replace(fishScriptCommand(), "");
      await fsAsync.writeFile(fishConfigPath, unboundFishConfig);
    }
  } catch {
    /* empty */
  }

  try {
    const powershellResolvedConfigPath = await powershellConfigPath();
    const powershellConfig = (await fsAsync.readFile(powershellResolvedConfigPath)).toString();
    if (powershellConfig.includes(powershellScriptCommand())) {
      const unboundPowershellConfig = powershellConfig.toString().replace(powershellScriptCommand(), "");
      await fsAsync.writeFile(powershellResolvedConfigPath, unboundPowershellConfig);
    }
  } catch {
    /* empty */
  }

  try {
    const pwshResolvedConfigPath = await pwshConfigPath();
    const pwshConfig = (await fsAsync.readFile(pwshResolvedConfigPath)).toString();
    if (pwshConfig.includes(pwshScriptCommand())) {
      const unboundPwshConfig = pwshConfig.toString().replace(pwshScriptCommand(), "");
      await fsAsync.writeFile(pwshResolvedConfigPath, unboundPwshConfig);
    }
  } catch {
    /* empty */
  }
};

export const deleteConfigFolder = async (): Promise<void> => {
  const cliConfigPath = path.join(os.homedir(), cacheFolder);
  if (fs.existsSync(cliConfigPath)) {
    fs.rmSync(cliConfigPath, { recursive: true });
  }
};

const safeAppendFile = async (filepath: string, data: string) => {
  if (!fs.existsSync(filepath)) {
    await fsAsync.mkdir(path.dirname(filepath), { recursive: true });
    await fsAsync.writeFile(filepath, data);
  } else {
    await fsAsync.appendFile(filepath, data);
  }
};

export const bind = async (shell: Shell): Promise<void> => {
  const cliConfigPath = path.join(os.homedir(), cacheFolder);
  if (!fs.existsSync(cliConfigPath)) {
    await fsAsync.mkdir(cliConfigPath);
  }
  switch (shell) {
    case Shell.Bash: {
      const bashConfigPath = path.join(os.homedir(), ".bashrc");
      await safeAppendFile(bashConfigPath, `\n${bashScriptCommand()}`);
      await fsAsync.copyFile(path.join(__dirname, "..", "..", "shell", "key-bindings.bash"), path.join(os.homedir(), cacheFolder, "key-bindings.bash"));
      break;
    }
    case Shell.Zsh: {
      const zshConfigPath = path.join(os.homedir(), ".zshrc");
      await safeAppendFile(zshConfigPath, `\n${zshScriptCommand()}`);
      await fsAsync.copyFile(path.join(__dirname, "..", "..", "shell", "key-bindings.zsh"), path.join(os.homedir(), cacheFolder, "key-bindings.zsh"));
      break;
    }
    case Shell.Fish: {
      const fishConfigPath = path.join(os.homedir(), ".config", "fish", "config.fish");
      await safeAppendFile(fishConfigPath, `\n${fishScriptCommand()}`);
      await fsAsync.copyFile(path.join(__dirname, "..", "..", "shell", "key-bindings.fish"), path.join(os.homedir(), cacheFolder, "key-bindings.fish"));
      break;
    }
    case Shell.Powershell: {
      await safeAppendFile(await powershellConfigPath(), `\n${powershellScriptCommand()}`);
      await fsAsync.copyFile(
        path.join(__dirname, "..", "..", "shell", "key-bindings-powershell.ps1"),
        path.join(os.homedir(), cacheFolder, "key-bindings-powershell.ps1"),
      );
      break;
    }
    case Shell.Pwsh: {
      await safeAppendFile(await pwshConfigPath(), `\n${pwshScriptCommand()}`);
      await fsAsync.copyFile(path.join(__dirname, "..", "..", "shell", "key-bindings-pwsh.ps1"), path.join(os.homedir(), cacheFolder, "key-bindings-pwsh.ps1"));
      break;
    }
  }
};
