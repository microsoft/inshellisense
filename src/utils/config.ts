// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import os from "node:os";
import path from "node:path";
import fs from "node:fs";
import fsAsync from "node:fs/promises";
import toml from "toml";
import _Ajv, { JSONSchemaType } from "ajv";
import { Command } from "commander";

const Ajv = _Ajv as unknown as typeof _Ajv.default;
const ajv = new Ajv();

export type PromptPattern = {
  regex: string;
  postfix: string;
};

type Binding = {
  shift?: boolean;
  control?: boolean;
  key: string;
};

type Config = {
  bindings: {
    nextSuggestion: Binding;
    previousSuggestion: Binding;
    dismissSuggestions: Binding;
    acceptSuggestion: Binding;
  };
  prompt?: {
    bash?: PromptPattern[];
    pwsh?: PromptPattern[];
    xonsh?: PromptPattern[];
    nu?: PromptPattern[];
    powershell?: PromptPattern[];
  };
  specs?: {
    path?: string[];
  };
};

const bindingSchema: JSONSchemaType<Binding> = {
  type: "object",
  nullable: true,
  properties: {
    shift: { type: "boolean", nullable: true },
    control: { type: "boolean", nullable: true },
    key: { type: "string" },
  },
  required: ["key"],
};

const promptPatternsSchema: JSONSchemaType<PromptPattern[]> = {
  type: "array",
  nullable: true,
  items: {
    type: "object",
    properties: {
      regex: { type: "string" },
      postfix: { type: "string" },
    },
    required: ["regex", "postfix"],
  },
};

const specPathsSchema: JSONSchemaType<string[]> = {
  type: "array",
  items: { type: "string" },
  nullable: true,
};

const configSchema = {
  type: "object",
  nullable: true,
  properties: {
    bindings: {
      type: "object",
      nullable: true,
      properties: {
        nextSuggestion: bindingSchema,
        previousSuggestion: bindingSchema,
        dismissSuggestions: bindingSchema,
        acceptSuggestion: bindingSchema,
      },
    },
    prompt: {
      type: "object",
      nullable: true,
      properties: {
        bash: promptPatternsSchema,
        pwsh: promptPatternsSchema,
        powershell: promptPatternsSchema,
        xonsh: promptPatternsSchema,
        nu: promptPatternsSchema,
      },
    },
    specs: {
      type: "object",
      nullable: true,
      properties: {
        path: specPathsSchema,
      },
    },
  },
  additionalProperties: false,
};

const configFile = ".inshellisenserc";
const cachePath = path.join(os.homedir(), ".inshellisense");
const configPath = path.join(os.homedir(), configFile);

let globalConfig: Config = {
  bindings: {
    nextSuggestion: { key: "down" },
    previousSuggestion: { key: "up" },
    acceptSuggestion: { key: "tab" },
    dismissSuggestions: { key: "escape" },
  },
};

export const getConfig = (): Config => globalConfig;
export const loadConfig = async (program: Command) => {
  if (fs.existsSync(configPath)) {
    let config: Config;
    try {
      config = toml.parse((await fsAsync.readFile(configPath)).toString());
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (e: any) {
      program.error(`${configFile} is invalid toml. Parsing error on line ${e.line}, column ${e.column}: ${e.message}`);
    }
    const isValid = ajv.validate(configSchema, config);
    if (!isValid) {
      program.error(`${configFile} is invalid: ${ajv.errorsText()}`);
    }
    globalConfig = {
      bindings: {
        nextSuggestion: config?.bindings?.nextSuggestion ?? globalConfig.bindings.nextSuggestion,
        previousSuggestion: config?.bindings?.previousSuggestion ?? globalConfig.bindings.previousSuggestion,
        acceptSuggestion: config?.bindings?.acceptSuggestion ?? globalConfig.bindings.acceptSuggestion,
        dismissSuggestions: config?.bindings?.dismissSuggestions ?? globalConfig.bindings.dismissSuggestions,
      },
      prompt: {
        bash: config.prompt?.bash,
        powershell: config.prompt?.powershell,
        xonsh: config.prompt?.xonsh,
        pwsh: config.prompt?.pwsh,
        nu: config.prompt?.nu,
      },
      specs: {
        path: [`${os.homedir()}/.fig/autocomplete/build`, ...(config?.specs?.path ?? [])],
      },
    };
  }
};

export const deleteCacheFolder = async (): Promise<void> => {
  const cliConfigPath = path.join(os.homedir(), cachePath);
  if (fs.existsSync(cliConfigPath)) {
    fs.rmSync(cliConfigPath, { recursive: true });
  }
};
