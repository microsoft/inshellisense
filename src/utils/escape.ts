// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

const ESC = "\u001B[";

const istermPromptStart = ESC + "?6973l";
const istermPromptEnd = ESC + "?69731l";
const cursorHide = ESC + "?25l";
const cursorShow = ESC + "?25h";

// \033]697;StartPrompt\007

export default { istermPromptEnd, istermPromptStart, cursorHide, cursorShow };
