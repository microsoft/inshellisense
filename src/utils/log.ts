// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import os from "node:os";
import path from "node:path";
import fs from "node:fs";
import fsAsync from "node:fs/promises";

const logTarget = path.join(os.homedir(), ".inshellisense", "inshellisense.log");
const logEnabled = false;

const reset = async () => {
  await fsAsync.writeFile(logTarget, "");
};

const debug = (content: object) => {
  if (!logEnabled) {
    return;
  }
  fs.appendFile(logTarget, JSON.stringify(content) + "\n", (err) => {
    if (err != null) {
      throw err;
    }
  });
};

export default { reset, debug };
