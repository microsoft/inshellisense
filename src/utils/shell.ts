// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import process from "node:process";
import find from "find-process";
import path from "node:path";
import which from "which";
import fs from "node:fs";
import url from "node:url";
import os from "node:os";
import fsAsync from "node:fs/promises";

export enum Shell {
  Bash = "bash",
  Powershell = "powershell",
  Pwsh = "pwsh",
  Zsh = "zsh",
  Fish = "fish",
  Cmd = "cmd",
}

export const supportedShells = [Shell.Bash, process.platform == "win32" ? Shell.Powershell : null, Shell.Pwsh, Shell.Zsh, Shell.Fish].filter(
  (shell) => shell != null,
) as Shell[];

export const userZdotdir = process.env?.ZDOTDIR ?? os.homedir() ?? `~`;
export const zdotdir = path.join(os.tmpdir(), `is-zsh`);

export const setupZshDotfiles = async () => {
  const shellFolderPath = path.join(path.dirname(url.fileURLToPath(import.meta.url)), "..", "..", "shell");
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-rc.zsh"), path.join(zdotdir, ".zshrc"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-profile.zsh"), path.join(zdotdir, ".zprofile"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-env.zsh"), path.join(zdotdir, ".zshenv"));
  await fsAsync.cp(path.join(shellFolderPath, "shellIntegration-login.zsh"), path.join(zdotdir, ".zlogin"));
};

export const inferShell = async () => {
  try {
    const name = path.parse(process.env.SHELL ?? "").name;
    const shellName = supportedShells.find((shell) => name.includes(shell));
    if (shellName) return shellName;
  } catch {
    /* empty */
  }
  const processResult = (await find("pid", process.ppid)).at(0);
  const name = processResult?.name;
  return name != null ? supportedShells.find((shell) => name.includes(shell)) : undefined;
};

export const gitBashPath = async (): Promise<string> => {
  const gitBashPaths = await getGitBashPaths();
  for (const gitBashPath of gitBashPaths) {
    if (fs.existsSync(gitBashPath)) {
      return gitBashPath;
    }
  }
  throw new Error("unable to find a git bash executable installed");
};

const getGitBashPaths = async (): Promise<string[]> => {
  const gitDirs: Set<string> = new Set();

  const gitExePath = await which("git.exe", { nothrow: true });
  if (gitExePath) {
    const gitExeDir = path.dirname(gitExePath);
    gitDirs.add(path.resolve(gitExeDir, "../.."));
  }

  const addValid = <T>(set: Set<T>, value: T | undefined) => {
    if (value) set.add(value);
  };

  // Add common git install locations
  addValid(gitDirs, process.env["ProgramW6432"]);
  addValid(gitDirs, process.env["ProgramFiles"]);
  addValid(gitDirs, process.env["ProgramFiles(X86)"]);
  addValid(gitDirs, `${process.env["LocalAppData"]}\\Program`);

  const gitBashPaths: string[] = [];
  for (const gitDir of gitDirs) {
    gitBashPaths.push(
      `${gitDir}\\Git\\bin\\bash.exe`,
      `${gitDir}\\Git\\usr\\bin\\bash.exe`,
      `${gitDir}\\usr\\bin\\bash.exe`, // using Git for Windows SDK
    );
  }

  // Add special installs that don't follow the standard directory structure
  gitBashPaths.push(`${process.env["UserProfile"]}\\scoop\\apps\\git\\current\\bin\\bash.exe`);
  gitBashPaths.push(`${process.env["UserProfile"]}\\scoop\\apps\\git-with-openssh\\current\\bin\\bash.exe`);

  return gitBashPaths;
};
