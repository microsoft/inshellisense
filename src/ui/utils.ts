// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import ansi from "ansi-escapes";
import wrapAnsi from "wrap-ansi";
import chalk from "chalk";
import wcwidth from "wcwidth";

/**
 * Renders a box around the given rows
 * @param rows the text content to be included in the box, must be <= width - 2
 * @param width the max width of a row
 * @param x the column to start the box at
 */
export const renderBox = (rows: string[], width: number, x: number, borderColor?: string) => {
  const result = [];
  const setColor = (text: string) => (borderColor ? chalk.hex(borderColor).apply(text) : text);
  result.push(ansi.cursorTo(x) + setColor("┌" + "─".repeat(width - 2) + "┐") + ansi.cursorTo(x));
  rows.forEach((row) => {
    result.push(ansi.cursorDown() + setColor("│") + row + setColor("│") + ansi.cursorTo(x));
  });
  result.push(ansi.cursorDown() + setColor("└" + "─".repeat(width - 2) + "┘") + ansi.cursorTo(x));
  return result.join("") + ansi.cursorUp(rows.length + 1);
};

export const truncateMultilineText = (description: string, width: number, maxHeight: number) => {
  const wrappedText = wrapAnsi(description, width, {
    trim: false,
    hard: true,
  });
  const lines = wrappedText.split("\n");
  const truncatedLines = lines.slice(0, maxHeight);
  if (lines.length > maxHeight) {
    truncatedLines[maxHeight - 1] = [...truncatedLines[maxHeight - 1]].slice(0, -1).join("") + "…";
  }
  return truncatedLines.map((line) => line.padEnd(width));
};

const wcPadEnd = (text: string, width: number, char = " "): string => text + char.repeat(Math.max(width - wcwidth(text), 0));

const wcPoints = (text: string, length: number): [string, boolean] => {
  const points = [...text];
  const accPoints = [];
  let accWidth = 0;
  for (const point of points) {
    const width = wcwidth(point);
    if (width + accWidth > length) {
      return wcwidth(accPoints.join("")) < length ? [accPoints.join(""), true] : [accPoints.slice(0, -1).join(""), true];
    }
    accPoints.push(point);
    accWidth += width;
  }
  return [accPoints.join(""), false];
};

/**
 * Truncates the text to the given width
 */
export const truncateText = (text: string, width: number) => {
  const [points, truncated] = wcPoints(text, width);
  return !truncated ? wcPadEnd(text, width) : wcPadEnd(points + "…", width);
};
