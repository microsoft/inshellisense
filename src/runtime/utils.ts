// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import path from "node:path";
import { spawn } from "node:child_process";
import fsAsync from "node:fs/promises";

import { CommandToken } from "./parser.js";
import { getPathSeperator, Shell } from "../utils/shell.js";
import log from "../utils/log.js";

export type ExecuteShellCommandTTYResult = {
  code: number | null;
};

export const buildExecuteShellCommand =
  (timeout: number): Fig.ExecuteCommandFunction =>
  async ({ command, env, args, cwd }: Fig.ExecuteCommandInput): Promise<Fig.ExecuteCommandOutput> => {
    const child = spawn(command, args, { cwd, env });
    setTimeout(() => child.kill("SIGKILL"), timeout);
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (data) => (stdout += data));
    child.stderr.on("data", (data) => (stderr += data));
    child.on("error", (err) => {
      log.debug({ msg: "shell command failed", e: err.message });
    });
    return new Promise((resolve) => {
      child.on("close", (code) => {
        resolve({
          status: code ?? 0,
          stderr,
          stdout,
        });
      });
    });
  };

export const resolveCwd = async (
  cmdToken: CommandToken | undefined,
  cwd: string,
  shell: Shell,
): Promise<{ cwd: string; pathy: boolean; complete: boolean }> => {
  if (cmdToken == null) return { cwd, pathy: false, complete: false };
  const { token } = cmdToken;
  const sep = getPathSeperator(shell);
  if (!token.includes(sep)) return { cwd, pathy: false, complete: false };
  const resolvedCwd = path.isAbsolute(token) ? token : path.join(cwd, token);
  try {
    await fsAsync.access(resolvedCwd, fsAsync.constants.R_OK);
    return { cwd: resolvedCwd, pathy: true, complete: token.endsWith(sep) };
  } catch {
    // fallback to the parent folder if possible
    const baselessCwd = resolvedCwd.substring(0, resolvedCwd.length - path.basename(resolvedCwd).length);
    try {
      await fsAsync.access(baselessCwd, fsAsync.constants.R_OK);
      return { cwd: baselessCwd, pathy: true, complete: token.endsWith(sep) };
    } catch {
      /*empty*/
    }
    return { cwd, pathy: false, complete: false };
  }
};
