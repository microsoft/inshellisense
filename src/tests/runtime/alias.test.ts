// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { jest } from "@jest/globals";
import { Shell } from "../../utils/shell";

const mockExecuteShellCommand = jest.fn();

jest.unstable_mockModule("../../runtime/utils.js", () => ({
  buildExecuteShellCommand: () => mockExecuteShellCommand,
}));

const { aliasExpand, loadAliases } = await import("../../runtime/alias.js");

beforeEach(() => {
  jest.clearAllMocks();
});

describe("aliasExpand", () => {
  test("expand on bash aliases", async () => {
    //@ts-expect-error - jest.fn() has no implementation
    mockExecuteShellCommand.mockResolvedValue({
      stdout: `alias glo='git log --oneline'
alias la='echo '\\''lo'\\'' '\\''la'\\'''
alias ls='ls --color=auto'`,
      status: 0,
    });

    await loadAliases(Shell.Bash);
    expect(aliasExpand([{ token: "glo", complete: false, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "la", complete: true, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "git", complete: true, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "ls", complete: true, isOption: false }])).toMatchSnapshot();
  });

  test("expand on zsh aliases", async () => {
    //@ts-expect-error - jest.fn() has no implementation
    mockExecuteShellCommand.mockResolvedValue({
      stdout: `glo='git log --oneline'
la='echo '\\''lo'\\'' '\\''la'\\'''
ls='ls --color=auto'`,
      status: 0,
    });

    await loadAliases(Shell.Zsh);
    expect(aliasExpand([{ token: "glo", complete: false, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "la", complete: true, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "git", complete: true, isOption: false }])).toMatchSnapshot();
    expect(aliasExpand([{ token: "ls", complete: true, isOption: false }])).toMatchSnapshot();
  });
});
