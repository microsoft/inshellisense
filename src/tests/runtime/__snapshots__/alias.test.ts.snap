// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aliasExpand expand on bash aliases 1`] = `
[
  {
    "complete": false,
    "isOption": false,
    "token": "glo",
  },
]
`;

exports[`aliasExpand expand on bash aliases 2`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "echo",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'lo'",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'la'",
  },
]
`;

exports[`aliasExpand expand on bash aliases 3`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "git",
  },
]
`;

exports[`aliasExpand expand on bash aliases 4`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "ls",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--color",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "auto",
  },
]
`;

exports[`aliasExpand expand on zsh aliases 1`] = `
[
  {
    "complete": false,
    "isOption": false,
    "token": "glo",
  },
]
`;

exports[`aliasExpand expand on zsh aliases 2`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "echo",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'lo'",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "'la'",
  },
]
`;

exports[`aliasExpand expand on zsh aliases 3`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "git",
  },
]
`;

exports[`aliasExpand expand on zsh aliases 4`] = `
[
  {
    "complete": true,
    "isOption": false,
    "token": "ls",
  },
  {
    "complete": true,
    "isOption": true,
    "token": "--color",
  },
  {
    "complete": true,
    "isOption": false,
    "token": "auto",
  },
]
`;
